cmd/theatre-envconsul/main.go

```diff
@@ -0,0 +1,314 @@
+package main
+
+import (
+	"context"
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"path"
+	"strings"
+	"syscall"
+
+	"k8s.io/client-go/rest"
+
+	"gopkg.in/yaml.v2"
+
+	"github.com/alecthomas/kingpin"
+	kitlog "github.com/go-kit/kit/log"
+	"github.com/hashicorp/vault/api"
+	"github.com/pkg/errors"
+
+	"github.com/gocardless/theatre/cmd"
+	"github.com/gocardless/theatre/pkg/signals"
+)
+
+var logger kitlog.Logger
+
+var (
+	app = kingpin.New("theatre-envconsul", "Kubernetes container vault support using envconsul").Version(Version)
+
+	commonOpts         = cmd.NewCommonOptions(app)
+	defaultInstallPath = "/var/theatre-vault"
+
+	install                       = app.Command("install", "Install binaries into path")
+	installPath                   = install.Flag("path", "Path to install theatre binaries").Default(defaultInstallPath).String()
+	installEnvconsulBinary        = install.Flag("envconsul-binary", "Path to envconsul binary").Default("/envconsul").String()
+	installTheatreEnvconsulBinary = install.Flag("theatre-envconsul-binary", "Path to theatre-envconsul binary").Default(os.Args[0]).String()
+
+	exec                     = app.Command("exec", "Authenticate with vault and exec envconsul")
+	execConfigFile           = exec.Flag("config-file", "app config file").String()
+	execAuthBackendMountPath = exec.Flag("auth-backend-mount-path", "Vault auth backend mount path").Default("kubernetes").String()
+	execAuthBackendRole      = exec.Flag("auth-backend-role", "Vault auth backend role").Default("default").String()
+	execCommand              = exec.Flag("command", "Command to execute").Required().String()
+	execInstallPath          = exec.Flag("install-path", "Path containing installed binaries").Default(defaultInstallPath).String()
+	execVaultAddress         = exec.Flag("vault-address", "Address of vault (format: scheme://host:port)").Required().String()
+
+	// Version is set at compile time
+	Version = "dev"
+)
+
+func main() {
+	command := kingpin.MustParse(app.Parse(os.Args[1:]))
+	logger = commonOpts.Logger()
+
+	ctx, cancel := signals.SetupSignalHandler()
+	defer cancel()
+
+	if err := mainError(ctx, command); err != nil {
+		logger.Log("error", err, "msg", "exiting with error")
+
+		os.Exit(1)
+	}
+}
+
+func mainError(ctx context.Context, command string) (err error) {
+	switch command {
+	case install.FullCommand():
+		files := []string{
+			*installEnvconsulBinary,
+			*installTheatreEnvconsulBinary,
+		}
+
+		logger.Log("msg", "copying files into install path", "file_path", *installPath)
+		for _, f := range files {
+			if err := copyExecutable(f, path.Join(*installPath, f)); err != nil {
+				return errors.Wrap(err, "error copying file")
+			}
+		}
+
+	case exec.FullCommand():
+		clusterConfig, err := rest.InClusterConfig()
```

Comment:
I think it' helpful to support running this outside of Kubernetes too. This gives a much quicker feedback loop, and has been what I've been using for testing.

The changes to allow this are pretty simple, e.g. (most of this is just whitespace changes)
```diff
diff --git cmd/theatre-envconsul/main.go cmd/theatre-envconsul/main.go
index 3ecd055..676dc91 100644
--- cmd/theatre-envconsul/main.go
+++ cmd/theatre-envconsul/main.go
@@ -46,6 +46,7 @@ var (
 	execCommand              = exec.Flag("command", "Command to execute").Required().String()
 	execInstallPath          = exec.Flag("install-path", "Path containing installed binaries").Default(defaultInstallPath).String()
 	execVaultAddress         = exec.Flag("vault-address", "Address of vault (format: scheme://host:port)").Required().String()
+	execVaultToken           = exec.Flag("vault-token", "Vault token to use, instead of Kubernetes auth").OverrideDefaultFromEnvar("VAULT_TOKEN").String()
 
 	// Version is set at compile time
 	Version = "dev"
@@ -81,15 +82,21 @@ func mainError(ctx context.Context, command string) (err error) {
 		}
 
 	case exec.FullCommand():
-		clusterConfig, err := rest.InClusterConfig()
-		if err != nil {
-			return errors.Wrap(err, "failed to authenticate within kubernetes")
-		}
+		var vaultToken string
+
+		if *execVaultToken != "" {
+			vaultToken = *execVaultToken
+		} else {
+			clusterConfig, err := rest.InClusterConfig()
+			if err != nil {
+				return errors.Wrap(err, "failed to authenticate within kubernetes")
+			}
 
-		logger.Log("event", "vault_login", "host", *execVaultAddress, "backend", *execAuthBackendMountPath, "role", *execAuthBackendRole)
-		vaultToken, err := getVaultToken(clusterConfig.BearerToken, *execVaultAddress, *execAuthBackendMountPath, *execAuthBackendRole)
-		if err != nil {
-			return errors.Wrap(err, "failed to login to vault")
+			logger.Log("event", "vault_login", "host", *execVaultAddress, "backend", *execAuthBackendMountPath, "role", *execAuthBackendRole)
+			vaultToken, err = getVaultToken(clusterConfig.BearerToken, *execVaultAddress, *execAuthBackendMountPath, *execAuthBackendRole)
+			if err != nil {
+				return errors.Wrap(err, "failed to login to vault")
+			}
 		}
 
 		logger.Log("event", "loading_config", "file_path", *execConfigFile)

```

Reply:
What token were you using for testing Ben? Your Vault account shouldn't permit retrieval of secrets, so it's a bit worrying if you were able to pull them!

Reply:
I was using a local vault server in dev mode

Reply:
Ah gotcha, thought this was lab for a minute